compiler/rustc_data_structures/src/xxhash/pointers.rs

```diff
@@ -0,0 +1,166 @@
+//! This module defines Ptr and PtrMut, which are pointers into slices that get their
+//! bounds checked only when debug assertions are enabled.
+
+// yuck.
+macro_rules! slice_to_ptr {
+    ($slice:expr, Ptr) => {
+        $slice.as_ptr()
+    };
+    ($slice:expr, PtrMut) => {
+        $slice.as_mut_ptr()
+    };
+}
+
+macro_rules! define_ptr {
+    ($name:ident, $ptr_ty:ty, $from_slice_ty:ty) => {
+        #[derive(Clone, Copy, PartialEq, Eq, Debug)]
```

Comment:
Is `PartialEq`/`Eq` needed? Seems a bit weird with the bounds. You could potentially have a `Ptr` to a slice, and then another `Ptr` to a subslice. The pointers themselves would be equal, but the bounds would not. Should they be considered equal? With the derive they would not be considered equal.